### PR TITLE
@ #423 | update travis-ci

### PR DESCRIPTION
**Pull request by hieptranquoc**
connect #423 
Please check this PR! Thank you.

### CODE REVIEW

File: spec/utility_spec.rb [hunk 1]
```diff
@@ -0,0 +1,416 @@
+require 'json'
```
Comment (hoatle) [inline spec/utility_spec.rb]: should mirror the `lib` directory => move the file to `spec/teracy-dev/util_spec.rb`

File: .travis.yml [hunk 2]
```diff
@@ -1,7 +1,31 @@
...
-services:
-  - docker
+git:
+  depth: 1
 
-script:
-- echo "OK"
+jobs:
+  include:
+    - stage: "Tests"
```
Comment (hoatle) [inline .travis.yml:13]: let's have these stages: "check style" => "test" => "deploy"

File: lib/teracy-dev/extension/manager.rb [hunk 3]
```diff
@@ -39,50 +38,50 @@ def install(extensions)
...
+        }
+        extension['location'].merge!(location)
         sync_existing = extension['path']['lookup'] == DEFAULT_EXTENSION_LOOKUP_PATH
 
         Location::Manager.sync(extension['location'], sync_existing)
-
       end
 
       def validate(extension)
-        return unless Util.true?(extension['enabled'])
```
Comment (hoatle) [inline lib/teracy-dev/extension/manager.rb]: should return immediately when the extension is not enabled

File: .rubocop.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
+AllCops:
```
Comment (hoatle) [inline .rubocop.yml:1]: follow our common standards here http://dev.teracy.org/docs/develop/style_standards.html#common-standards

File: lib/teracy-dev/extension/manager.rb [hunk 5]
```diff
@@ -94,13 +93,13 @@ def validate_dependencies(extension, extention_manifest_list)
...
             abort
-          else
-            if !Util.require_version_valid?(found['version'], dependency['require_version'])
-              @logger.error("the extension #{dependency['name']} #{dependency['require_version']} is required by "\
-              "#{extension_manifest['name']} but found version #{found['version']}")
-              abort
-            end
           end
+
+          next if Util.require_version_valid?(found['version'], dependency['require_version'])
```
Comment (hoatle) [inline lib/teracy-dev/extension/manager.rb]: return if

File: lib/teracy-dev/location/git_synch.rb [hunk 6]
```diff
@@ -1,9 +1,8 @@
 require 'open3'
-
```
Comment (hoatle) [inline lib/teracy-dev/location/git_synch.rb:2]: need @datphan to check on this

File: lib/teracy-dev/logging/progname_acceptor.rb [hunk 7]
```diff
@@ -2,19 +2,18 @@
 
 module TeracyDev
   module Logging
-    # filter messages by LOG_PROGNAME env var
-    # LOG_PROGNAME=TeracyDev => display all logs has progname matching TeracyDev, eg: TeracyDev, TeracyDev::Config, etc
-    # LOG_PROGNAME=TeracyDev::Config => display all logs has progname matching TeracyDev::Config, eg: TeracyDev::Config, TeracyDev::Config::Provisioners, etc
-    # LOG_PROGNAME="TeracyDev::Loader|TeracyDev::Location" => display all logs having progname matching this regex
+    # filter messages by log_progname env var
```
Comment (hoatle) [inline lib/teracy-dev/logging/progname_acceptor.rb]: LOG_PROGNAME must be kept here in the docs

File: lib/teracy-dev/util.rb [hunk 8]
```diff
@@ -87,67 +84,67 @@ def self.load_yaml_file(file_path)
...
-      @@logger.debug("version: #{version.strip}; requirements: #{requirements.strip}")
+      @logger.debug("version: #{version.strip}; requirements: #{requirements.strip}")
       requirements = requirements.split(',')
       req = Gem::Requirement.new(*requirements)
       req.satisfied_by?(Gem::Version.new(version))
     end
 
     # teracy-dev hash override algorithm
-    # override the originalHash object with the sourceHash object
+    # override the originalHash object with the source_hash object
```
Comment (hoatle) [inline lib/teracy-dev/util.rb]: origin_hash

File: spec/fixture/config.json [hunk 9]
```diff
@@ -0,0 +1,162 @@
+{
```
Comment (hoatle) [inline spec/fixture/config.json]: move the util fixtures to `spec/fixtures/teracy-dev/util` directory instead

**Review by hoatle — CHANGES_REQUESTED** [review]
@hieptranquoc please check my comments
need @phuonglm and @datphan to continue reviewing

File: lib/teracy-dev/location/git_synch.rb [hunk 10]
```diff
@@ -133,29 +129,25 @@ def start(location, sync_existing)
...
-              elsif tag
-                updated = check_tag(current_ref, tag)
-              else
-                updated = check_branch(current_ref, branch)
-              end
+              updated = check_ref(current_ref, ref) if ref
+
+              updated = check_tag(current_ref, tag) if tag
+
+              updated = check_branch(current_ref, branch)
```
Comment (datphan) [inline lib/teracy-dev/location/git_synch.rb]: I don't think these lines is correct logic

Comment (datphan) [reply lib/teracy-dev/location/git_synch.rb]: exp: if `tag` is present then `check_branch` is still get called

File: lib/teracy-dev/location/git_synch.rb [hunk 11]
```diff
@@ -53,15 +52,12 @@ def initialize
...
-        begin
-          start(location_conf, sync_existing)
-        rescue GitWarn => e
-          @logger.warn(e)
-
-          false
-        rescue
-          raise
-        end
+        start(location_conf, sync_existing)
```
Comment (datphan) [inline lib/teracy-dev/location/git_synch.rb:62]: it should have `begin`

**Comment by hieptranquoc** [comment]
I applied all comment, please check this PR again! Thank you all.

File: lib/teracy-dev/location/git_synch.rb [hunk 12]
```diff
@@ -408,52 +418,51 @@ def get_remote_credentials remote_url
...
+        credential_exists = !ENV[repo_username_key].nil? && !ENV[repo_password_key].nil?
 
-        @logger.debug("repo_username: #{repo_username_key}=#{ENV[repo_username_key]}, repo_password_key: #{repo_password_key}=#{ENV[repo_password_key]}")
+        @logger.debug("repo_username: #{repo_username_key}=#{ENV[repo_username_key]}, " \
+        "repo_password_key: #{repo_password_key}=#{ENV[repo_password_key]}")
 
         # only if user defined credentials are not present
         if credential_exists
-          credential_str = "'#{CGI.escape ENV[repo_username_key]}':'#{CGI.escape ENV[repo_password_key]}'"
+          credential_str = "'#{CGI.escape ENV[repo_username_key]}': " \
```
Comment (datphan) [inline lib/teracy-dev/location/git_synch.rb]: it should not have `': "` (space) character

File: lib/teracy-dev/location/git_synch.rb [hunk 13]
```diff
@@ -126,36 +132,33 @@ def start(location, sync_existing)
...
-              elsif tag
-                updated = check_tag(current_ref, tag)
-              else
-                updated = check_branch(current_ref, branch)
-              end
+              updated = check_ref(current_ref, ref) if ref
+
+              updated = check_tag(current_ref, tag) if tag
+
+              updated ||= check_branch(current_ref, branch)
```
Comment (datphan) [inline lib/teracy-dev/location/git_synch.rb]: if `check_tag` or `check_ref` return `false` then `check_branch` will still get called

File: lib/teracy-dev/util.rb [hunk 14]
```diff
@@ -24,23 +22,26 @@ def self.exist?(value)
...
     end
 
-
     # convert a value (case insentitive) to a boolen value
     # true, "true", "t", "yes", "y", "1" => true
     # othwerwise, false
     #
     # @since v0.6.0-a5
     def self.true?(value)
-      begin
```
Comment (datphan) [inline lib/teracy-dev/util.rb]: should have `begin`

Comment (hieptranquoc) [reply lib/teracy-dev/util.rb]: > should have `begin`

In Ruby 2.5 or later, you can omit `begin` in `do-end` block.

Comment (hieptranquoc) [reply lib/teracy-dev/util.rb]: we will change this in the future!

**Comment by hieptranquoc** [comment]
I've updated! Please check again!

File: lib/teracy-dev/util.rb (hunk 14 above)
Comment (datphan) [reply lib/teracy-dev/util.rb]: So we need to force user to use ruby >=2.5 which is not necessary, my ruby version:
```
$ ruby -v
ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-darwin16]
```

Comment (hoatle) [reply lib/teracy-dev/util.rb]: @datphan make sure to review the latest changes from the PR instead.

**Comment by hieptranquoc** [comment]
@hoatle please check this PR!

**Comment by hoatle** [comment]
need @datphan to approve this.

File: lib/teracy-dev/extension/manager.rb [hunk 15]
```diff
@@ -26,34 +27,42 @@ def install(extensions)
           end
         end
 
-        extention_manifest_list = extensions.map { |x| Manager.manifest(x) if Util.true?(x['enabled'])}.compact
+        ext_manifest_list = extensions.map { |x| Manager.manifest(x) if Util.true?(x['enabled']) }
+        # returns a copy of self with all nil elements removed.
+        ext_manifest_list.compact
```
Comment (hoatle) [inline lib/teracy-dev/extension/manager.rb]: `compact!` or `ext_manifest_list = ext_manifest_list.compact`

**Review by hoatle — CHANGES_REQUESTED** [review]

**Comment by hoatle** [comment]
lots of changes, so we need to review and test this intensively before merging. still waiting for @datphan to approve this.

**Review by datphan — DISMISSED** [review]

**Review by hoatle — APPROVED** [review]
LGTM, will test more to fix bugs if any.

**Pull request merged**